src/unix/netmap-udp.c

```diff
@@ -0,0 +1,735 @@
+/* Copyright Joyent, Inc. and other Node contributors. All rights reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#include "uv.h"
+#include "internal.h"
+
+#include <assert.h>
+#include <string.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <unistd.h>
+#if defined(__MVS__)
+#include <xti.h>
+#endif
+#include <sys/un.h>
+
+#define NETMAP_WITH_LIBS
+#include <net/netmap_user.h>
+#include <net/ethernet.h>
+#include <netinet/ip.h>
+#include <netinet/udp.h>
+
+
+static uint16_t ip_checksum(struct ip* ip);
+static void uv__udp_netmap_recv_packet(uv_loop_t* loop, struct netmap_slot* slot, uint8_t* p);
+static size_t uv__udp_netmap_generate_udp(uv_loop_t* loop, unsigned int src_port, struct msghdr* h, uint8_t* pkt);
+static int uv__udp_netmap_send_udp(uv_loop_t* loop, uv_udp_send_t* req, struct netmap_ring* ring);
+static void uv__udp_netmap_run_completed(uv_loop_t* loop);
+static void uv__udp_netmap_io(uv_loop_t* loop, uv__io_t* w, unsigned int revents);
+
+static const size_t ETH_LEN = sizeof(struct ether_header);
+static const size_t ETH_IP_LEN = ETH_LEN + sizeof(struct ip);
+static const size_t ETH_IP_UDP_LEN = ETH_IP_LEN + sizeof(struct udphdr);
+
+
+static uint16_t ip_checksum(struct ip* ip) {
```

Comment:
Do you need to bail out `if (ip == NULL)`?

Reply:
Since these checksum methods are `static` I'm ok with possibly derefing a NULL ptr. I also poked around some in udp.c and in general libuv doesn't seem to do a lot of NULL checking.